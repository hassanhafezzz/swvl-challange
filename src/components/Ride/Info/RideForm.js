import React, { useContext } from 'react';
import { Formik, Form } from 'formik';
import classNames from 'classnames/bind';
import styles from './styles.module.css';
import {
  LATENCY_EARLY,
  LATENCY_ON_TIME,
  LATENCY_LATE,
} from '../../../constants';

import { startTrip } from '../../../store/actions';
import { Context } from '../../../store';
import InputField from '../../common/Form/InputField';
import SelectField from '../../common/Form/SelectField';
import Button, { BUTTON_VARIANT } from '../../common/Button';

const cx = classNames.bind(styles);

const RideForm = ({ closeModalForm }) => {
  const initialValues = {
    duration: 3,
    latency: LATENCY_ON_TIME,
  };

  const [_, dispatch] = useContext(Context);

  const onSubmit = (values) => {
    let { duration } = values;
    const { latency } = values;

    switch (latency) {
      case LATENCY_EARLY:
        duration -= 0.5;
        break;
      case LATENCY_LATE:
        duration += 0.5;
        break;
      default:
        break;
    }
    dispatch(startTrip(duration));
    closeModalForm();
  };

  const validate = (values) => {
    const errors = {};
    if (!values.duration) {
      errors.duration = 'What is the expected time for the trip?';
    }

    if (!values.latency) {
      errors.latency = 'Is the driver running late?';
    }

    return errors;
  };

  return (
    <Formik
      initialValues={initialValues}
      validate={validate}
      onSubmit={onSubmit}
    >
      {({ isValid }) => (
        <Form>
          <InputField
            type="number"
            name="duration"
            id="duration"
            label="Trip duration in minutes"
          />
          <SelectField name="latency" id="latency" label="Expected latency">
            <option disabled selected label="Is the driver running late?" />
            <option label="arriving early" value={LATENCY_EARLY} />
            <option label="arriving on time" value={LATENCY_ON_TIME} />
            <option label="arriving late" value={LATENCY_LATE} />
          </SelectField>

          <Button
            className={cx('start-button')}
            variant={BUTTON_VARIANT.SECONDARY}
            disabled={!isValid}
            type="submit"
          >
            Yalla
          </Button>
        </Form>
      )}
    </Formik>
  );
};

export default RideForm;
