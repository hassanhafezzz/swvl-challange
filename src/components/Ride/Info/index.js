import React, { useState } from 'react';
import classNames from 'classnames/bind';
import Button, { BUTTON_VARIANT } from '../../common/Button';
import Modal from '../../common/Modal';
import { ReactComponent as Star } from '../../../img/star.svg';
import { ReactComponent as Dollar } from '../../../img/dollar.svg';
import { ReactComponent as Pin } from '../../../img/pin.svg';
import driver from '../../../img/driver.jpg';
import bus from '../../../img/ford-bus.png';

import { getFormattedDate, getFormattedTime } from '../../../utils';

import styles from './styles.module.css';

const cx = classNames.bind(styles);

const Info = () => {
  const [isModalOpen, setModalOpen] = useState(false);
  const openModal = () => setModalOpen(true);
  const closeModal = () => setModalOpen(false);

  const now = new Date();
  return (
    <>
      <Modal title="Start ride ‚ö°" isOpen={isModalOpen} closeModal={closeModal}>
        placeholder for form
      </Modal>

      <div className={cx('container')}>
        <div className={cx('heading')}>
          <div>
            <h3 className={cx('title')}>
              Trip Information{' '}
              <span role="img" aria-label="bus">
                üöê
              </span>
            </h3>
            <p className={cx('time')}>
              <span>{getFormattedDate(now)}, </span>
              <span>{getFormattedTime(now)}</span>
            </p>
          </div>

          <Button variant={BUTTON_VARIANT.PRIMARY} onClick={openModal}>
            Start Ride
          </Button>
        </div>
        <div className={cx('widgets')}>
          <div className={cx('driver')}>
            <div className={cx('media')}>
              <img
                className={cx('driver-image')}
                src={driver}
                alt="driver saaid"
              />
              <img className={cx('bus-image')} src={bus} alt="ford minibus" />
            </div>
            <div className={cx('body')}>
              <p className={cx('driver-name')}>
                Herbert Patton
                <span className={cx('rating-icon')}>
                  <Star title="star" /> 5.0
                </span>
              </p>
              <p className={cx('bus-type')}>Ford Transit - GA 213</p>
            </div>
          </div>
          <div className={cx('route')}>
            <div className={cx('route-desc')}>
              <p>Almaza</p>
              <p>Talat harb axis</p>
            </div>
          </div>
          <div className={cx('extra-info')}>
            <p>
              <Pin />
              Trip Distance: 23 KM
            </p>
            <p>
              <Dollar />
              Trip Base Fare: 45 EGP
            </p>
          </div>
        </div>
      </div>
    </>
  );
};

export default Info;
