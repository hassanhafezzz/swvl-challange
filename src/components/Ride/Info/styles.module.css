@value utils: '../../../styles/utils.css';
@value large-bp from utils;
@value medium-bp from utils;

.container {
  margin: 20px 0;
}

.heading {
  margin: 15px 0 10px;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.title {
  margin: 5px 0;
  font-size: var(--h1-size);
}

.time {
  margin: 0;
  color: var(--color-gray-300);
}

.widgets {
  margin-top: 15px;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.widgets > div {
  margin: 20px 0;
}

.driver {
  display: flex;
  align-items: center;
}

.media {
  display: flex;
  align-items: center;
  justify-content: center;
}

.driver-image {
  z-index: 99;
  width: 70px;
  height: 70px;
  border-radius: 50%;
  box-shadow: 2px 2px 15px -2px var(--color-gray-300);
}

.bus-image {
  margin-left: -25px;
  height: 70px;
}

.body {
  margin-left: 10px;
}

.driver-name {
  font-weight: var(--font-bold);
  margin: 5px 0;
}

.rating-icon {
  color: var(--color-gray-300);
  font-weight: var(--font-normal);
  margin: 0 5px;
}

.rating-icon svg {
  width: 13px;
  height: 13px;
  fill: var(--color-gray-300);
}

.bus-type {
  margin: 5px 0;
  font-size: var(--small-paragraph);
}

.route {
  display: flex;
}

.route::before {
  content: '';
  display: flex;
  width: 5px;
  height: auto;
  background: linear-gradient(
    0deg,
    var(--color-light-secondary) 0%,
    var(--color-primary) 100%
  );
  border-radius: 4px;
}

.route-desc {
  margin-left: 10px;
}

.route-desc p {
  font-weight: var(--font-bold);
  margin: 0;
}

.route-desc p + p {
  margin-top: 10px;
}

.extra-info {
  display: flex;
  flex-direction: column;
}

.extra-info p {
  margin: 0;
  display: flex;
}

.extra-info p + p {
  margin-top: 10px;
}

.extra-info svg {
  padding-top: 2px;
  margin-right: 5px;
  height: 14px;
  width: 14px;
}

@media (medium-bp) {
  .widgets {
    flex-direction: column;
    align-items: flex-start;
  }
}
