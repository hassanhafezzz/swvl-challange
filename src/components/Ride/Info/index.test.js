import React from 'react';
import { render } from '@testing-library/react';
import StoreProvider from '../../../store';
import Info from './index';

const ele = (
  <StoreProvider>
    <Info />
  </StoreProvider>
);

describe('info component test', () => {
  test('renders Info section', () => {
    const { getByTestId } = render(ele);
    const info = getByTestId('info');
    expect(info).toBeTruthy();
  });

  test('start ride button exist', () => {
    const { container, getByText } = render(ele);
    const buttons = container.querySelectorAll('button');
    expect(buttons.length).toEqual(1);
    expect(getByText(/start ride/i)).toBeTruthy();
  });

  test('add snapshot for info', () => {
    const { container, asFragment } = render(ele);
    expect(asFragment(container)).toMatchSnapshot();
  });
});
