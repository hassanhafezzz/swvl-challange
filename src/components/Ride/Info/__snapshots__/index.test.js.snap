// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`info component test add snapshot for info 1`] = `
<DocumentFragment>
  <section
    class="container"
    data-testid="info"
  >
    <div
      class="heading"
    >
      <div>
        <h3
          class="title"
        >
          Trip Information 
          <span
            aria-label="bus"
            role="img"
          >
            üöê
          </span>
        </h3>
      </div>
      <button
        class="button secondary"
        disabled=""
        type="button"
      >
        Start Ride
      </button>
    </div>
    <div
      class="widgets"
    >
      <div
        class="driver"
      >
        <div
          class="media"
        >
          <img
            alt="Herbert Patton"
            class="driver-image"
            src="driver.jpg"
          />
          <img
            alt="ford minibus"
            class="bus-image"
            src="ford-bus.png"
          />
        </div>
        <div
          class="body"
        >
          <p
            class="driver-name"
          >
            Herbert Patton
            <span
              class="rating-icon"
            >
              <svg
                title="star"
              >
                star.svg
              </svg>
               5
            </span>
          </p>
          <p
            class="bus-type"
          >
            Ford Transit - GA 213
          </p>
        </div>
      </div>
      <div
        class="route"
      >
        <div
          class="route-desc"
        >
          <p>
            almaza
          </p>
          <p>
            talat harb
          </p>
        </div>
      </div>
      <div
        class="extra-info"
      >
        <p>
          <svg>
            pin.svg
          </svg>
          Trip Distance: 13.798 KM
        </p>
        <p>
          <svg>
            dollar.svg
          </svg>
          Trip Base Fare: 55 EGP
        </p>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
